core/src/main/java/com/google/cloud/sql/ApplicationDefaultCredentialFactory.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql;
+
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.services.sqladmin.SQLAdminScopes;
+import com.google.auth.http.HttpCredentialsAdapter;
+import com.google.auth.oauth2.GoogleCredentials;
+import java.io.IOException;
+import java.util.Arrays;
+
+public class ApplicationDefaultCredentialFactory implements CredentialFactory {
+
+  @Override
+  public HttpRequestInitializer create() {
+    GoogleCredentials credentials;
+    try {
+      credentials = GoogleCredentials.getApplicationDefault();
+    } catch (IOException err) {
+      throw new RuntimeException(
+          "Unable to obtain credentials to communicate with the Cloud SQL API", err);
+    }
+    if (credentials.createScopedRequired()) {
```

Comment:
What does this method do? Shouldn't we always be adding OAuth2 scopes? Maybe this is a quirk of the Java API?

Reply:
Yeah, it's a java quirk. There's a chance we could have credentials here that are already scoped, and then the call to createScoped would fail. See 
https://cloud.google.com/java/docs/reference/google-auth-library/latest/com.google.auth.oauth2.GoogleCredentials#com_google_auth_oauth2_GoogleCredentials_createScopedRequired__